source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc

```diff
@@ -208,17 +210,15 @@ void GradientController::recordLatencySample(MonotonicTime rq_send_time) {
   const std::chrono::microseconds rq_latency =
       std::chrono::duration_cast<std::chrono::microseconds>(time_source_.monotonicTime() -
                                                             rq_send_time);
-  uint32_t sample_count;
   {
     absl::MutexLock ml(&sample_mutation_mtx_);
     hist_insert(latency_sample_hist_.get(), rq_latency.count(), 1);
-    sample_count = hist_sample_count(latency_sample_hist_.get());
   }
 
-  if (inMinRTTSamplingWindow() && sample_count >= config_.minRTTAggregateRequestCount()) {
-    // This sample has pushed the request count over the request count requirement for the minRTT
-    // recalculation. It must now be finished.
+  synchronizer_.syncPoint("pre_minrtt_update");
```

Comment:
my bad, didn't mention it early about the reason why doing it in this way:
As we are using the single mutex for both [updateMinRTT()](https://github.com/envoyproxy/envoy/blob/9711a5d3f40c6e3faa7627c58a85893e1e19d830/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L219-L222) and [hist_insert(..)](https://github.com/envoyproxy/envoy/blob/9711a5d3f40c6e3faa7627c58a85893e1e19d830/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L213-L216), If adding one sync point inside updateMinRTT(), it will cause another worker thread to pend on the hist_insert.

Reply:
You need to block both threads at the first sync point so that this doesn't happen.

Reply:
Okay, just added another syncPoint `post_minrtt_update` to test the behavior that one thread won't invoke updateMinRTT() as it is not able to acquire the mutex. 

Tried adding another test against the case about entering minrtt update window, but found `advanceTimeAndRun` doesn't work well in multiple-threads env, seems to align with [the description](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/test/test_common/test_time_system.h#L72) 
 
- if invoking `advanceTimeAndRun` inside thread t2, it will cause assertion failure.

-  if invoking `advanceTimeAndRun` in the thread on the context of this test, the enableTimer invoked in t2 seems not effective.

Reply:
You don't need to have this logic outside of the main test thread. You can just pause the worker threads at the sync points with barriers and then call the advance time.

Reply:
yeah, this is the second case I mentioned above, but unclear to me why it doesn't work.